src/copy.cpp

```diff
@@ -0,0 +1,98 @@
+#include <filesystem>
+#include <fstream>
+#include <vector>
+#include <iostream>
+#include <thread>
+#include <mutex>
+#include <copy.h>
+#include <exceptions/io_exception.h>
+
+std::exception_ptr Copy::exceptionPtr = nullptr;
+
+Copy::Copy(CopyArgumentParser args) : _args(std::move(args)), _done_reading(false) {
+    std::cout << "Try to copy " << _args.sourcePath() << " to " << _args.destinationPath() << std::endl;
+}
+
+void Copy::runOneThread(const int buffer_size) const {
+
+    std::ifstream fin(_args.sourcePath(), std::ifstream::binary);
+    if (!fin.is_open())
+        throw CopyFileReadException(_args.sourcePath());
+
+    std::ofstream fout(_args.destinationPath(), std::ifstream::binary);
+    if (!fout.is_open())
+        throw CopyFileWriteException(_args.destinationPath());
+
+    std::unique_ptr<char[]> buffer(new char[buffer_size]);
+
+    while (fin.good()) {
+        fin.read(buffer.get(), buffer_size);
+        std::streamsize s = fin.gcount();
+
+        if (fout.good())
+            fout.write(buffer.get(), s);
+        else
+            throw CopyFileWriteException(_args.sourcePath());
+    }
+}
+
+void Copy::parallelCopy(const int buffer_size) {
+    std::thread reader(&Copy::reader, this, buffer_size);
```

Comment:
There is no need to make a _done_reading and the queue as a class member.
You could declare here a local bool object, and pass it to both threads.
With an existing design, it's not possible to use this class twice for copying 2 times files (which is bad).